topologySpreadConstriants with empty labelSelector makes karpenter crash

## Version
<!-- https://github.com/aws/karpenter/releases -->
**Karpenter**: v0.5.3
<!-- kubectl version | grep Server -->
**Kubernetes**: v1.22.4

## Actual Behavior
<!-- Briefly describe what is actually happening -->

Karpenter crash with an npr if a Pod has a topologySpreadConstraint with empty labelSelector. Empty labelSelectors doesn't make much sense, but still shouldn't crash karpenter.

## Resource Specs and Logs

Just ignoring the spread constraint is probably fine if the labelSelector is empty.

```
2021-12-19T07:40:14.064Z        INFO    controller.provisioning Batched 2 pods in 1.000147818s  {"commit": "5047f3c", "provisioner": "karpenter-nodes"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x19daf66]

goroutine 10410 [running]:
github.com/aws/karpenter/pkg/controllers/provisioning/scheduling.TopologyListOptions({0xc000bdb0b0, 0xb}, 0xc00410e820)
        github.com/aws/karpenter/pkg/controllers/provisioning/scheduling/topology.go:144 +0x66
github.com/aws/karpenter/pkg/controllers/provisioning/scheduling.(*Topology).countMatchingPods(0xc00376cf10, {0x27965b8, 0xc001184db0}, 0xc00410e820)
        github.com/aws/karpenter/pkg/controllers/provisioning/scheduling/topology.go:122 +0x85
github.com/aws/karpenter/pkg/controllers/provisioning/scheduling.(*Topology).computeZonalTopology(0xc00410e870, {0x27965b8, 0xc001184db0}, {0xc0024acea0, 0x0, 0x0}, 0x0)
        github.com/aws/karpenter/pkg/controllers/provisioning/scheduling/topology.go:114 +0x7e
github.com/aws/karpenter/pkg/controllers/provisioning/scheduling.(*Topology).computeCurrentTopology(0x0, {0x27965b8, 0xc001184db0}, 0xc000c5dda0, 0xc00410e820)
        github.com/aws/karpenter/pkg/controllers/provisioning/scheduling/topology.go:82 +0x93
github.com/aws/karpenter/pkg/controllers/provisioning/scheduling.(*Topology).Inject(0xc00118e2d8, {0x27965b8, 0xc001184db0}, 0xc000c5dda0, {0xc004902790, 0xc000c5dd68, 0x1})
        github.com/aws/karpenter/pkg/controllers/provisioning/scheduling/topology.go:45 +0xbf
github.com/aws/karpenter/pkg/controllers/provisioning/scheduling.(*Scheduler).Solve(0xc003769740, {0x27965b8, 0xc001184db0}, 0xc00118e1c0, {0xc004902790, 0x2, 0x2})
        github.com/aws/karpenter/pkg/controllers/provisioning/scheduling/scheduler.go:73 +0x1c8
github.com/aws/karpenter/pkg/controllers/provisioning.(*Provisioner).provision(0xc0003cc540, {0x27965b8, 0xc001184db0})
        github.com/aws/karpenter/pkg/controllers/provisioning/provisioner.go:116 +0x208
github.com/aws/karpenter/pkg/controllers/provisioning.NewProvisioner.func1()
        github.com/aws/karpenter/pkg/controllers/provisioning/provisioner.go:65 +0x73
created by github.com/aws/karpenter/pkg/controllers/provisioning.NewProvisioner
        github.com/aws/karpenter/pkg/controllers/provisioning/provisioner.go:63 +0x3a5
```

I also saw crash with below configs
      nodeSelector:
        kubernetes.io/arch: amd64    
      topologySpreadConstraints:
      - maxSkew: 5
        topologyKey: "topology.kubernetes.io/zone"
        whenUnsatisfiable: ScheduleAnyway

2021-12-21T11:47:25.368Z        INFO    controller.provisioning Batched 18 pods in 1.506678526s {"commit": "5047f3c", "provisioner": "default"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x19daf66]

goroutine 133726 [running]:
github.com/aws/karpenter/pkg/controllers/provisioning/scheduling.TopologyListOptions({0xc006bbf4e0, 0x7}, 0xc0046b1630)
        github.com/aws/karpenter/pkg/controllers/provisioning/scheduling/topology.go:144 +0x66
github.com/aws/karpenter/pkg/controllers/provisioning/scheduling.(*Topology).countMatchingPods(0xc003d2ea80, {0x27965b8, 0xc0046af710}, 0xc0046b1630)
        github.com/aws/karpenter/pkg/controllers/provisioning/scheduling/topology.go:122 +0x85
github.com/aws/karpenter/pkg/controllers/provisioning/scheduling.(*Topology).computeZonalTopology(0x0, {0x27965b8, 0xc0046af710}, {0xc0044ff6c0, 0x0, 0x0}, 0x0)
        github.com/aws/karpenter/pkg/controllers/provisioning/scheduling/topology.go:114 +0x7e
github.com/aws/karpenter/pkg/controllers/provisioning/scheduling.(*Topology).computeCurrentTopology(0x0, {0x27965b8, 0xc0046af710}, 0xc004737da0, 0xc0046b1630)
        github.com/aws/karpenter/pkg/controllers/provisioning/scheduling/topology.go:82 +0x93
github.com/aws/karpenter/pkg/controllers/provisioning/scheduling.(*Topology).Inject(0xc003c070d8, {0x27965b8, 0xc0046af710}, 0xc004737da0, {0xc0046b5800, 0xc004737d68, 0xc0045e6320})
        github.com/aws/karpenter/pkg/controllers/provisioning/scheduling/topology.go:45 +0xbf
github.com/aws/karpenter/pkg/controllers/provisioning/scheduling.(*Scheduler).Solve(0xc0001830e0, {0x27965b8, 0xc0046af710}, 0xc003c06fc0, {0xc0046b5800, 0x12, 0x20})
        github.com/aws/karpenter/pkg/controllers/provisioning/scheduling/scheduler.go:73 +0x1c8
github.com/aws/karpenter/pkg/controllers/provisioning.(*Provisioner).provision(0xc002f60cb0, {0x27965b8, 0xc0046af710})
        github.com/aws/karpenter/pkg/controllers/provisioning/provisioner.go:116 +0x208
github.com/aws/karpenter/pkg/controllers/provisioning.NewProvisioner.func1()
        github.com/aws/karpenter/pkg/controllers/provisioning/provisioner.go:65 +0x73
created by github.com/aws/karpenter/pkg/controllers/provisioning.NewProvisioner
        github.com/aws/karpenter/pkg/controllers/provisioning/provisioner.go:63 +0x3a5

--- a/pkg/controllers/provisioning/scheduling/suite_test.go
+++ b/pkg/controllers/provisioning/scheduling/suite_test.go
@@ -493,7 +493,7 @@ var _ = Describe("Topology", func() {
 			Expect(env.Client.List(ctx, &nodes)).To(Succeed())
 			ExpectSkew(ctx, env.Client, "default", &topology[0]).To(ConsistOf(2, 2, 1))
 		})
-		It("should match all pods when labelSelector is nil", func() {
+		It("should match all pods when labelSelector is not specified", func() {
 			topology := []v1.TopologySpreadConstraint{{
 				TopologyKey:       v1.LabelTopologyZone,
 				WhenUnsatisfiable: v1.DoNotSchedule,

--- a/pkg/controllers/provisioning/scheduling/suite_test.go
+++ b/pkg/controllers/provisioning/scheduling/suite_test.go
@@ -493,6 +493,18 @@ var _ = Describe("Topology", func() {
 			Expect(env.Client.List(ctx, &nodes)).To(Succeed())
 			ExpectSkew(ctx, env.Client, "default", &topology[0]).To(ConsistOf(2, 2, 1))
 		})
+		It("should match all pods when labelSelector is nil", func() {
+			topology := []v1.TopologySpreadConstraint{{
+				TopologyKey:       v1.LabelTopologyZone,
+				WhenUnsatisfiable: v1.DoNotSchedule,
+				LabelSelector:     nil,
+				MaxSkew:           1,
+			}}
+			ExpectProvisioned(ctx, env.Client, selectionController, provisioners, provisioner,
+				test.UnschedulablePod(test.PodOptions{ObjectMeta: metav1.ObjectMeta{Labels: labels}}),
+			)
+			ExpectSkew(ctx, env.Client, "default", &topology[0]).To(ConsistOf(1))
+		})
 	})
 
 	Context("Hostname", func() {

--- a/pkg/controllers/provisioning/scheduling/suite_test.go
+++ b/pkg/controllers/provisioning/scheduling/suite_test.go
@@ -497,11 +497,10 @@ var _ = Describe("Topology", func() {
 			topology := []v1.TopologySpreadConstraint{{
 				TopologyKey:       v1.LabelTopologyZone,
 				WhenUnsatisfiable: v1.DoNotSchedule,
-				LabelSelector:     nil,
 				MaxSkew:           1,
 			}}
 			ExpectProvisioned(ctx, env.Client, selectionController, provisioners, provisioner,
-				test.UnschedulablePod(test.PodOptions{ObjectMeta: metav1.ObjectMeta{Labels: labels}}),
+				test.UnschedulablePod(),
 			)
 			ExpectSkew(ctx, env.Client, "default", &topology[0]).To(ConsistOf(1))
 		})

--- a/pkg/controllers/provisioning/scheduling/topology.go
+++ b/pkg/controllers/provisioning/scheduling/topology.go
@@ -141,6 +141,9 @@ func (t *Topology) countMatchingPods(ctx context.Context, topologyGroup *Topolog
 
 func TopologyListOptions(namespace string, constraint *v1.TopologySpreadConstraint) *client.ListOptions {
 	selector := labels.Everything()
+	if constraint.LabelSelector == nil {
+		return &client.ListOptions{Namespace: namespace, LabelSelector: selector}
+	}
 	for key, value := range constraint.LabelSelector.MatchLabels {
 		requirement, _ := labels.NewRequirement(key, selection.Equals, []string{value})
 		selector = selector.Add(*requirement)

